docs/markdown/Rewriter.md

```diff
@@ -0,0 +1,220 @@
+---
+short-description: Automatic modification of the build system files
+...
+
+# Meson file rewriter
+
+Since version 0.50.0, meson has the functionality to perform some basic
+modification on the `meson.build` files from the command line. The currently
+supported operations are:
+
+- For build targets:
+  - Add/Remove source files
+  - Add/Remove targets
+  - Modify a select set of kwargs
+  - Print some JSON information
+- For dependencies:
+  - Modify a select set of kwargs
+- For the project function:
+  - Modify a select set of kwargs
+  - Modify the default options list
+
+The rewriter has both, a normal command line interface and a "script mode". The
+normal CLI is mostly designed for everyday use. The "script mode", on the
+other hand, is meant to be used by external programs (IDEs, graphical
+frontends, etc.)
+
+## Using the rewriter
+
+All rewriter functions are accessed via `meson rewrite`. The meson rewriter
+assumes that it is run inside the project root directory. If this isn't the
+case, use `--sourcedir` to specify the actual project source directory.
+
+### Adding and removing sources
+
+The most common operations will probably be the adding and removing of source
+files to a build target. This can be easily done with:
+
+```bash
+meson rewrite target <target name/id> {add/rm} [list of sources]
+```
+
+For instance, given the following example
+
+```meson
+src = ['main.cpp', 'fileA.cpp']
+
+exe1 = executable('testExe', src)
+```
+
+the source `fileB.cpp` can be added with:
+
+```bash
+meson rewrite target testExe add fileB.cpp
+```
+
+After executing this command, the new `meson.build` will look like this:
+
+```meson
+src = ['main.cpp', 'fileA.cpp', 'fileB.cpp']
+
+exe1 = executable('testExe', src)
+```
+
+In this case, `exe1` could also have been used for the target name, since the
+rewriter also takes assignments and internal meson IDs into consideration when
+searching for the target in the `meson.build` files.
+
+For more information see the help output of the rewriter target command.
```

Comment:
There should be some explanation here about how to deal with the case that a target with the same name is defined in two different directories (and also how subprojects tie into this).

Reply:
Is it even possible in meson to have two targets with the same name? If yes, then the first target that matches either the ID or the target name would be selected.

Reply:
Yes it is. Blindly selecting the first can lead to confusing behaviour. If there are multiple targets with the same name it should error out and print something like:

```
Target name "foo" is ambiguous. Operation not performed. To do this use the id instead of name. The potential choices are:

foo@asdf1234 in subdir/abc
foo@qqqq0000 in foo/bar/baz
...
```

Reply:
In that case, I will also add that. I had always assumed that target names have to be unique.

Reply:
Done (even with a test case)

Reply:
Done